File: backend/routers/material_master.py
"""자재마스터 조회 API."""
from fastapi import APIRouter, Query, HTTPException
from typing import Optional
from database.connection import get_connection

router = APIRouter(prefix="/api/material-master", tags=["material-master"])

# 시스템에서 노출할 자재유형 (그 외는 모두 숨김)
ALLOWED_TYPES = ('ERSA', 'HIBE', 'ERSB')


@router.get("/search")
def search(
    q: Optional[str] = Query(None, description="자재번호/내역/제조자PN 검색"),
    product_group_prefix: Optional[str] = Query(None, description="용품별그룹 앞 4자리(BB01 등)"),
    material_type: Optional[str] = None,
    is_unused: Optional[bool] = None,
    limit: int = 50,
    offset: int = 0,
):
    """자재마스터 검색. ALLOWED_TYPES(ERSA/HIBE/ERSB)만 노출."""
    conn = get_connection()
    try:
        where = []
        params = []
        if q and q.strip():
            qs = q.strip()
            where.append("(material_no ILIKE ? OR material_desc ILIKE ? OR manufacturer_pn ILIKE ? OR legacy_material_no ILIKE ?)")
            like = f'%{qs}%'
            params += [like, like, like, like]
        if product_group_prefix:
            where.append("product_group LIKE ?")
            params.append(f'{product_group_prefix}%')
        if material_type and material_type in ALLOWED_TYPES:
            where.append("material_type = ?")
            params.append(material_type)
        else:
            # 유형 미선택 시에도 ALLOWED_TYPES만 노출
            placeholders = ','.join(['?'] * len(ALLOWED_TYPES))
            where.append(f"material_type IN ({placeholders})")
            params += list(ALLOWED_TYPES)
        if is_unused is not None:
            where.append("is_unused = ?")
            params.append(is_unused)

        where_clause = ('WHERE ' + ' AND '.join(where)) if where else ''
        sql = f"""
            SELECT material_no, material_desc, manufacturer_pn, unit,
                   product_group, product_group_desc,
                   material_group, material_group_desc,
                   material_type, material_type_desc,
                   importance, importance_desc,
                   procurement, procurement_desc,
                   lead_time_days, legacy_material_no, created_date,
                   is_unused
            FROM material_master
            {where_clause}
            ORDER BY material_no
            LIMIT ? OFFSET ?
        """
        params_with_paging = params + [limit, offset]
        rows = conn.execute(sql, tuple(params_with_paging)).fetchall()
        items = [dict(r) for r in rows]

        # BOM 연결 정보 일괄 조회
        if items:
            mat_nos = [it['material_no'] for it in items]
            placeholders = ','.join(['?'] * len(mat_nos))
            bom_rows = conn.execute(f"""
                SELECT b.corp_material_no, b.id, b.material_no AS bom_code, b.name,
                       vt.code AS vehicle_code, vt.name AS vehicle_name
                FROM bom_nodes b LEFT JOIN vehicle_types vt ON vt.id = b.vehicle_type_id
                WHERE b.corp_material_no IN ({placeholders})
            """, tuple(mat_nos)).fetchall()
            bom_map = {}
            for r in bom_rows:
                bom_map.setdefault(r['corp_material_no'], []).append({
                    'id': r['id'], 'bom_code': r['bom_code'], 'name': r['name'],
                    'vehicle_code': r['vehicle_code'], 'vehicle_name': r['vehicle_name'],
                })
            for it in items:
                it['bom_links'] = bom_map.get(it['material_no'], [])

        count_sql = f"SELECT COUNT(*) AS cnt FROM material_master {where_clause}"
        total = conn.execute(count_sql, tuple(params)).fetchone()['cnt']
        return {
            'items': items,
            'total': total,
            'limit': limit,
            'offset': offset,
        }
    finally:
        conn.close()


@router.get("/groups")
def groups():
    """용품별그룹 4자리 분류 + 자재유형."""
    conn = get_connection()
    try:
        # 용품별그룹 앞 4자리(BB01 등)로 묶기, 알파벳순 정렬
        product_group_prefixes = conn.execute("""
            SELECT LEFT(product_group, 4) AS prefix,
                   COUNT(*) AS cnt,
                   COUNT(DISTINCT product_group) AS sub_count
            FROM material_master
            WHERE product_group ~ '^[A-Z]{2}[0-9]{2}'
            GROUP BY prefix
            ORDER BY prefix
        """).fetchall()
        placeholders = ','.join(['?'] * len(ALLOWED_TYPES))
        material_types = conn.execute(f"""
            SELECT material_type AS code, material_type_desc AS name, COUNT(*) AS cnt
            FROM material_master
            WHERE material_type IN ({placeholders})
            GROUP BY material_type, material_type_desc
            ORDER BY cnt DESC
        """, tuple(ALLOWED_TYPES)).fetchall()
        return {
            'product_group_prefixes': [dict(r) for r in product_group_prefixes],
            'material_types': [dict(r) for r in material_types],
        }
    finally:
        conn.close()


@router.get("/stats")
def stats():
    """전체 통계 (ALLOWED_TYPES만)."""
    conn = get_connection()
    try:
        placeholders = ','.join(['?'] * len(ALLOWED_TYPES))
        row = conn.execute(f"""
            SELECT
                COUNT(*) AS total,
                COUNT(*) FILTER (WHERE material_type = 'ERSA') AS ersa_count,
                COUNT(*) FILTER (WHERE material_type = 'HIBE') AS hibe_count,
                COUNT(*) FILTER (WHERE material_type = 'ERSB') AS ersb_count,
                COUNT(*) FILTER (WHERE is_unused = TRUE) AS unused_count,
                COUNT(*) FILTER (WHERE is_unused = FALSE OR is_unused IS NULL) AS active_count
            FROM material_master
            WHERE material_type IN ({placeholders})
        """, tuple(ALLOWED_TYPES)).fetchone()
        return dict(row)
    finally:
        conn.close()


@router.get("/{material_no}")
def detail(material_no: str):
    """자재 상세."""
    conn = get_connection()
    try:
        row = conn.execute("""
            SELECT * FROM material_master WHERE material_no = ?
        """, (material_no,)).fetchone()
        if not row:
            raise HTTPException(404, "자재를 찾을 수 없습니다")
        m = dict(row)
        bom_links = conn.execute("""
            SELECT b.id, b.material_no AS bom_code, b.name, b.vehicle_type_id,
                   vt.code AS vehicle_code, vt.name AS vehicle_name
            FROM bom_nodes b LEFT JOIN vehicle_types vt ON vt.id = b.vehicle_type_id
            WHERE b.corp_material_no = ?
        """, (material_no,)).fetchall()
        m['bom_links'] = [dict(r) for r in bom_links]
        return m
    finally:
        conn.close()

File: frontend/src/api/materialMaster.ts
import client from './client'

export interface BomLink {
  id: number
  bom_code: string | null
  name: string
  vehicle_code: string | null
  vehicle_name: string | null
}

export interface MaterialMasterItem {
  material_no: string
  material_desc: string | null
  manufacturer_pn: string | null
  unit: string | null
  product_group: string | null
  product_group_desc: string | null
  material_group: string | null
  material_group_desc: string | null
  material_type: string | null
  material_type_desc: string | null
  importance: string | null
  importance_desc: string | null
  procurement: string | null
  procurement_desc: string | null
  lead_time_days: number | null
  legacy_material_no: string | null
  created_date: string | null
  is_unused: boolean | null
  bom_links?: BomLink[]
}

export interface SearchResult {
  items: MaterialMasterItem[]
  total: number
  limit: number
  offset: number
}

export interface PrefixItem { prefix: string; cnt: number; sub_count: number }
export interface TypeItem   { code: string; name: string; cnt: number }

export interface Groups {
  product_group_prefixes: PrefixItem[]
  material_types: TypeItem[]
}

export interface Stats {
  total: number
  ersa_count: number
  hibe_count: number
  unused_count: number
  active_count: number
}

export interface MaterialDetail extends MaterialMasterItem {
  spec: string | null
  type_approval_desc: string | null
  safety_cert_desc: string | null
  safety_type_desc: string | null
  industry_desc: string | null
  inspection_desc: string | null
  eval_class_desc: string | null
  bom_links: BomLink[]
}

export interface EcatAttribute { name: string; value: string; unit: string; desc: string }
export interface EcatImage { filename: string; url: string }
export interface EcatMaterial {
  material_no: string
  material_desc_full: string
  material_desc_40: string
  group_classification_no: string
  group_classification_name: string
  group_classification_en: string
  group_classification_desc: string
  product_group: string
  product_group_desc: string
  unit: string
  manufacturer_pn: string
  created_date: string
  procurement_name: string
  safety_code: string
  safety_type: string
  is_unused: boolean
  use_plant: string
  stock_count: string
  optimal_stock: string
  lead_time: string
  yearly_plan_amount: string
  registered_company_count: string
  contract_in_progress: string
  contract_completed: string
  avg_contract_amount: string
  category_2: string
  purchase_group: string
  purchase_group_name: string
  last_update_date: string
  last_update_time: string
  attributes: EcatAttribute[]
  images: EcatImage[]
  source_url: string
}

export interface SyncResult {
  scanned_range: string
  scanned_count: number
  found_count: number
  next_start: number
}

export const ecatApi = {
  material: (materialNo: string): Promise<EcatMaterial> =>
    client.get(`/ecat/material/${encodeURIComponent(materialNo)}`).then(r => r.data),

  syncNew: (params: { start?: number; count?: number; concurrency?: number }): Promise<SyncResult> =>
    client.post('/ecat/sync-new', params).then(r => r.data),

  syncStatus: (): Promise<{ total: number; max_no: number; recent_updated: number }> =>
    client.get('/ecat/sync-status').then(r => r.data),
}

export const materialMasterApi = {
  search: (params: {
    q?: string
    product_group_prefix?: string
    material_type?: string
    is_unused?: boolean
    limit?: number
    offset?: number
  }): Promise<SearchResult> =>
    client.get('/material-master/search', { params }).then(r => r.data),

  groups: (): Promise<Groups> =>
    client.get('/material-master/groups').then(r => r.data),

  stats: (): Promise<Stats> =>
    client.get('/material-master/stats').then(r => r.data),

  detail: (materialNo: string): Promise<MaterialDetail> =>
    client.get(`/material-master/${encodeURIComponent(materialNo)}`).then(r => r.data),
}

// 용품별그룹 prefix 라벨 (코레일 표 기준)
// 자재번호 prefix 별 큰 분류:
//   1xxxxxx (ERSA) 보수품  → AA~FF + II
//   7xxxxxx (HIBE) 비재고품 → GG, JJ, KK (+ 일부)
export const PRODUCT_GROUP_PREFIX_LABEL: Record<string, string> = {
  // 고속차량
  'AA01': '궤도',
  // 일반차량 (BB01~BB04)
  'BB01': 'KTX',
  'BB02': '디젤동차',
  'BB03': '디젤기관차',
  'BB04': '전기기관차',
  // 전동차량 (BB05~BB08)
  'BB05': '전동차',
  'BB06': '객차',
  'BB07': '화차',
  'BB08': '발전차',
  // 시설/전기/기타 (BB09~BB13)
  'BB09': '차량복구장비',
  'BB10': '차량공통',
  'BB11': '간선형전기동차',
  'BB12': 'KTX-산천',
  'BB13': 'KTX-이음',
  // 전철 및 전력 / 통신 / 신호 (CC01~CC04)
  'CC01': '전철 및 전력',
  'CC02': '통신',
  'CC03': '신호',
  'CC04': '전기/신호 공통',
  // 금속 / 시설 / 연료
  'DD01': '금속',
  'EE01': '시설',
  'FF01': '연료 및 윤활류',
  // 기계기구·보선·대표 (보수품/비)
  'GG01': '기계기구및비품',
  'HH01': '제용품',
  'II01': '보선장비',
  'JJ00': '대표자재(기계기구)',
  'KK00': '대표자재(제용품)',
}

// prefix별 큰 구분 (보수품/비)
export const PRODUCT_GROUP_CATEGORY: Record<string, '보수품' | '비' | '보수품·비'> = {
  AA01: '보수품',
  BB01: '보수품', BB02: '보수품', BB03: '보수품', BB04: '보수품',
  BB05: '보수품', BB06: '보수품', BB07: '보수품', BB08: '보수품',
  BB09: '보수품', BB10: '보수품', BB11: '보수품', BB12: '보수품', BB13: '보수품',
  CC01: '보수품', CC02: '보수품', CC03: '보수품', CC04: '보수품',
  DD01: '보수품',
  EE01: '보수품',
  FF01: '비',
  GG01: '보수품·비',
  HH01: '비',
  II01: '보수품',
  JJ00: '비',
  KK00: '비',
}

// 색상으로 묶이는 큰 분류 (코레일 표의 우측 색 범례 기준)
// 고속차량 / 일반차량 / 전동차량 / 시설 / 전기 / 기타
export const PRODUCT_GROUP_BIG_CATEGORY: Record<string, '고속차량' | '일반차량' | '전동차량' | '시설' | '전기' | '기타'> = {
  AA01: '고속차량',
  BB01: '일반차량', BB02: '일반차량', BB03: '일반차량', BB04: '일반차량',
  BB05: '전동차량', BB06: '전동차량', BB07: '전동차량', BB08: '전동차량',
  BB09: '시설', BB10: '전기', BB11: '전동차량', BB12: '일반차량', BB13: '일반차량',
  CC01: '전기', CC02: '전기', CC03: '전기', CC04: '전기',
  DD01: '기타',
  EE01: '시설',
  FF01: '시설',
  GG01: '기타',
  HH01: '기타',
  II01: '시설',
  JJ00: '기타',
  KK00: '기타',
}

export const BIG_CATEGORY_COLOR: Record<string, string> = {
  '고속차량': '#8c8c8c',
  '일반차량': '#1677ff',
  '전동차량': '#722ed1',
  '시설':     '#cf1322',
  '전기':     '#52c41a',
  '기타':     '#bfbfbf',
}
